File: automation/tests/saleads_mi_negocio_full_test.py
#!/usr/bin/env python3
"""
End-to-end browser test for the SaleADS "Mi Negocio" workflow.

The flow is environment agnostic:
- no hardcoded domain
- prefers visible text selectors
- supports either an already-opened browser page (via CDP) or a provided start URL
"""

from __future__ import annotations

import argparse
import json
import re
import sys
import time
from dataclasses import dataclass
from datetime import datetime
from pathlib import Path
from typing import Callable

from playwright.sync_api import Locator
from playwright.sync_api import Page
from playwright.sync_api import Playwright
from playwright.sync_api import TimeoutError as PlaywrightTimeoutError
from playwright.sync_api import sync_playwright


REPORT_FIELDS = [
    "Login",
    "Mi Negocio menu",
    "Agregar Negocio modal",
    "Administrar Negocios view",
    "Información General",
    "Detalles de la Cuenta",
    "Tus Negocios",
    "Términos y Condiciones",
    "Política de Privacidad",
]


@dataclass
class StepResult:
    status: str
    details: list[str]


class SaleadsMiNegocioWorkflow:
    def __init__(self, page: Page, artifacts_dir: Path, account_email: str) -> None:
        self.page = page
        self.account_email = account_email
        self.artifacts_dir = artifacts_dir
        self.results: dict[str, StepResult] = {
            key: StepResult(status="FAIL", details=[]) for key in REPORT_FIELDS
        }
        self.legal_urls: dict[str, str] = {}
        self.screenshots: list[str] = []
        self.screenshot_index = 1

    def _log(self, message: str) -> None:
        print(f"[saleads-mi-negocio] {message}", flush=True)

    def _sanitize(self, value: str) -> str:
        safe = re.sub(r"[^a-zA-Z0-9_-]+", "_", value.strip().lower())
        return safe.strip("_") or "checkpoint"

    def wait_for_ui(self, page: Page | None = None) -> None:
        current_page = page or self.page
        try:
            current_page.wait_for_load_state("domcontentloaded", timeout=15000)
        except PlaywrightTimeoutError:
            pass
        try:
            current_page.wait_for_load_state("networkidle", timeout=15000)
        except PlaywrightTimeoutError:
            pass
        current_page.wait_for_timeout(600)

    def take_screenshot(self, name: str, full_page: bool = False, page: Page | None = None) -> str:
        current_page = page or self.page
        filename = f"{self.screenshot_index:02d}_{self._sanitize(name)}.png"
        self.screenshot_index += 1
        path = self.artifacts_dir / filename
        current_page.screenshot(path=str(path), full_page=full_page)
        self.screenshots.append(str(path))
        self._log(f"Screenshot saved: {path}")
        return str(path)

    def _locators_for_text(self, text: str, page: Page | None = None) -> list[Locator]:
        current_page = page or self.page
        escaped = re.escape(text)
        text_pattern = re.compile(escaped, re.IGNORECASE)
        return [
            current_page.get_by_role("button", name=text_pattern),
            current_page.get_by_role("link", name=text_pattern),
            current_page.get_by_role("menuitem", name=text_pattern),
            current_page.get_by_role("tab", name=text_pattern),
            current_page.get_by_text(text_pattern),
        ]

    def click_by_visible_text(
        self, text_candidates: list[str], wait_after_click: bool = True, page: Page | None = None
    ) -> str:
        current_page = page or self.page
        last_error: Exception | None = None

        for text in text_candidates:
            for locator in self._locators_for_text(text, current_page):
                target = locator.first
                try:
                    target.wait_for(state="visible", timeout=2500)
                    target.click()
                    if wait_after_click:
                        self.wait_for_ui(current_page)
                    self._log(f"Clicked text target: {text}")
                    return text
                except Exception as exc:  # noqa: BLE001
                    last_error = exc

        raise AssertionError(
            f"Unable to click any candidate text: {text_candidates}. Last error: {last_error}"
        )

    def _is_locator_visible(self, locator: Locator, timeout_ms: int = 5000) -> bool:
        try:
            locator.first.wait_for(state="visible", timeout=timeout_ms)
            return True
        except Exception:  # noqa: BLE001
            return False

    def is_any_text_visible(self, text_candidates: list[str], timeout_ms: int = 6000, page: Page | None = None) -> bool:
        current_page = page or self.page
        deadline = time.time() + (timeout_ms / 1000.0)
        while time.time() < deadline:
            for text in text_candidates:
                for locator in self._locators_for_text(text, current_page):
                    if self._is_locator_visible(locator, timeout_ms=400):
                        return True
            time.sleep(0.2)
        return False

    def assert_text_visible(self, text_candidates: list[str], description: str, page: Page | None = None) -> None:
        if not self.is_any_text_visible(text_candidates, page=page):
            raise AssertionError(f"Expected visible text not found for {description}: {text_candidates}")

    def run_step(self, label: str, action: Callable[[], None]) -> None:
        self._log(f"Running step: {label}")
        try:
            action()
            self.results[label].status = "PASS"
        except Exception as exc:  # noqa: BLE001
            self.results[label].status = "FAIL"
            self.results[label].details.append(str(exc))
            self._log(f"Step failed [{label}]: {exc}")
            try:
                self.take_screenshot(f"{label}_failure", full_page=True)
            except Exception as screenshot_exc:  # noqa: BLE001
                self._log(f"Failure screenshot could not be captured: {screenshot_exc}")

    def _maybe_click_google_account(self, popup_page: Page) -> None:
        self.wait_for_ui(popup_page)
        if self.is_any_text_visible([self.account_email], page=popup_page):
            self.click_by_visible_text([self.account_email], wait_after_click=True, page=popup_page)
            return

        # Fallback for account chooser with strict equality spacing differences.
        candidate = popup_page.get_by_text(re.compile(re.escape(self.account_email), re.IGNORECASE)).first
        if self._is_locator_visible(candidate, timeout_ms=4000):
            candidate.click()
            self.wait_for_ui(popup_page)

    def step_login(self) -> None:
        context = self.page.context
        previous_pages = set(context.pages)
        self.click_by_visible_text(
            [
                "Sign in with Google",
                "Iniciar sesión con Google",
                "Iniciar sesion con Google",
                "Continuar con Google",
                "Ingresar con Google",
                "Login with Google",
            ],
            wait_after_click=False,
        )

        popup_page: Page | None = None
        deadline = time.time() + 8
        while time.time() < deadline:
            current_pages = set(context.pages)
            new_pages = [candidate for candidate in current_pages if candidate not in previous_pages]
            if new_pages:
                popup_page = new_pages[0]
                break
            time.sleep(0.2)

        if popup_page is not None:
            self._log("Google login popup detected")
            self._maybe_click_google_account(popup_page)
            # Wait for popup to close or finish redirect.
            try:
                popup_page.wait_for_event("close", timeout=15000)
            except PlaywrightTimeoutError:
                pass

        # If account chooser appears in same tab, try selecting the requested account.
        if self.is_any_text_visible([self.account_email]):
            self.click_by_visible_text([self.account_email])

        self.wait_for_ui(self.page)

        main_app_visible = (
            self._is_locator_visible(self.page.locator("main"), timeout_ms=7000)
            or self._is_locator_visible(self.page.locator("aside"), timeout_ms=7000)
            or self._is_locator_visible(self.page.get_by_role("navigation"), timeout_ms=7000)
        )
        if not main_app_visible:
            raise AssertionError("Main application interface was not detected after login")

        sidebar_visible = (
            self._is_locator_visible(self.page.locator("aside"), timeout_ms=7000)
            or self._is_locator_visible(self.page.get_by_role("navigation"), timeout_ms=7000)
            or self.is_any_text_visible(["Negocio", "Mi Negocio"], timeout_ms=7000)
        )
        if not sidebar_visible:
            raise AssertionError("Left sidebar navigation was not detected after login")

        self.take_screenshot("dashboard_loaded")

    def step_mi_negocio_menu(self) -> None:
        if self.is_any_text_visible(["Negocio"], timeout_ms=3000):
            self.click_by_visible_text(["Negocio"])

        self.click_by_visible_text(["Mi Negocio", "Mi negocio"])

        self.assert_text_visible(["Agregar Negocio"], "Mi Negocio expanded submenu item Agregar Negocio")
        self.assert_text_visible(
            ["Administrar Negocios", "Administrar negocios"],
            "Mi Negocio expanded submenu item Administrar Negocios",
        )
        self.take_screenshot("mi_negocio_menu_expanded")

    def step_agregar_negocio_modal(self) -> None:
        self.click_by_visible_text(["Agregar Negocio"])

        self.assert_text_visible(["Crear Nuevo Negocio"], "Agregar Negocio modal title")
        self.assert_text_visible(["Nombre del Negocio"], "Agregar Negocio input label")
        self.assert_text_visible(
            ["Tienes 2 de 3 negocios", "Tienes 2 de 3 negocio"],
            "Agregar Negocio quota text",
        )
        self.assert_text_visible(["Cancelar"], "Agregar Negocio modal cancel button")
        self.assert_text_visible(["Crear Negocio"], "Agregar Negocio modal create button")

        self.take_screenshot("agregar_negocio_modal")

        name_input_candidates = [
            self.page.get_by_label("Nombre del Negocio", exact=False),
            self.page.get_by_placeholder("Nombre del Negocio"),
            self.page.locator("input[name*='negocio' i], input[id*='negocio' i]"),
        ]
        for locator in name_input_candidates:
            if self._is_locator_visible(locator, timeout_ms=1500):
                locator.first.click()
                locator.first.fill("Negocio Prueba Automatización")
                break

        self.click_by_visible_text(["Cancelar"])

    def step_administrar_negocios_view(self) -> None:
        if not self.is_any_text_visible(["Administrar Negocios"], timeout_ms=2500):
            self.click_by_visible_text(["Mi Negocio", "Mi negocio"])

        self.click_by_visible_text(["Administrar Negocios", "Administrar negocios"])
        self.wait_for_ui()

        self.assert_text_visible(["Información General", "Informacion General"], "account section Información General")
        self.assert_text_visible(["Detalles de la Cuenta", "Detalles de la cuenta"], "account section Detalles")
        self.assert_text_visible(["Tus Negocios", "Tus negocios"], "account section Tus Negocios")
        self.assert_text_visible(["Sección Legal", "Seccion Legal", "Legal"], "account section legal")
        self.take_screenshot("administrar_negocios_account_page", full_page=True)

    def step_informacion_general(self) -> None:
        self.assert_text_visible(["Información General", "Informacion General"], "Información General heading")
        self.assert_text_visible(["BUSINESS PLAN"], "Business plan label")
        self.assert_text_visible(["Cambiar Plan", "Change Plan"], "Cambiar Plan button")

        body_text = self.page.locator("body").inner_text()
        has_email = bool(re.search(r"[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}", body_text))
        if not has_email:
            raise AssertionError("No visible user email detected in Información General")

        has_name_label = self.is_any_text_visible(["Nombre", "Usuario", "User"])
        has_person_name = bool(re.search(r"\b[A-ZÁÉÍÓÚÑ][a-záéíóúñ]+(?:\s+[A-ZÁÉÍÓÚÑ][a-záéíóúñ]+)+\b", body_text))
        if not (has_name_label or has_person_name):
            raise AssertionError("No visible user name signal detected in Información General")

    def step_detalles_cuenta(self) -> None:
        self.assert_text_visible(["Detalles de la Cuenta", "Detalles de la cuenta"], "Detalles heading")
        self.assert_text_visible(["Cuenta creada"], "Cuenta creada text")
        self.assert_text_visible(["Estado activo", "Activo"], "Estado activo text")
        self.assert_text_visible(["Idioma seleccionado", "Idioma"], "Idioma seleccionado text")

    def step_tus_negocios(self) -> None:
        self.assert_text_visible(["Tus Negocios", "Tus negocios"], "Tus Negocios heading")
        self.assert_text_visible(["Agregar Negocio"], "Tus Negocios add button")
        self.assert_text_visible(["Tienes 2 de 3 negocios"], "Tus Negocios quota")

        container = self.page.locator("section,div").filter(
            has_text=re.compile(r"Tus\s+Negocios", re.IGNORECASE)
        ).first
        if not self._is_locator_visible(container, timeout_ms=3000):
            raise AssertionError("Tus Negocios container is not visible")

        list_items = container.locator("li, [role='listitem'], table tbody tr, article, .card")
        if list_items.count() > 0:
            return

        # Generic fallback: ensure section has non-trivial content beyond static labels.
        section_text = container.inner_text().strip()
        if len(section_text) < 70:
            raise AssertionError("Business list content not detected in Tus Negocios section")

    def _open_legal_link_and_validate(self, link_text: str, heading_text: str, report_label: str) -> None:
        context = self.page.context
        previous_pages = set(context.pages)
        self.click_by_visible_text([link_text], wait_after_click=False)

        target_page = self.page
        deadline = time.time() + 8
        while time.time() < deadline:
            current_pages = set(context.pages)
            new_pages = [candidate for candidate in current_pages if candidate not in previous_pages]
            if new_pages:
                target_page = new_pages[0]
                break
            time.sleep(0.2)

        self.wait_for_ui(target_page)
        self.assert_text_visible([heading_text], f"{report_label} heading", page=target_page)

        page_text = target_page.locator("body").inner_text().strip()
        if len(page_text) < 200:
            raise AssertionError(f"{report_label} legal content seems too short")

        self.take_screenshot(self._sanitize(report_label), page=target_page)
        self.legal_urls[report_label] = target_page.url

        if target_page != self.page:
            target_page.close()
            self.page.bring_to_front()
            self.wait_for_ui(self.page)
        else:
            # Same-tab navigation fallback.
            self.page.go_back(wait_until="domcontentloaded")
            self.wait_for_ui(self.page)

    def step_terminos(self) -> None:
        self._open_legal_link_and_validate(
            link_text="Términos y Condiciones",
            heading_text="Términos y Condiciones",
            report_label="Términos y Condiciones",
        )

    def step_politica(self) -> None:
        self._open_legal_link_and_validate(
            link_text="Política de Privacidad",
            heading_text="Política de Privacidad",
            report_label="Política de Privacidad",
        )

    def run_all(self) -> dict[str, object]:
        self.run_step("Login", self.step_login)
        self.run_step("Mi Negocio menu", self.step_mi_negocio_menu)
        self.run_step("Agregar Negocio modal", self.step_agregar_negocio_modal)
        self.run_step("Administrar Negocios view", self.step_administrar_negocios_view)
        self.run_step("Información General", self.step_informacion_general)
        self.run_step("Detalles de la Cuenta", self.step_detalles_cuenta)
        self.run_step("Tus Negocios", self.step_tus_negocios)
        self.run_step("Términos y Condiciones", self.step_terminos)
        self.run_step("Política de Privacidad", self.step_politica)

        report = {
            "test_name": "saleads_mi_negocio_full_test",
            "generated_at_utc": datetime.utcnow().isoformat() + "Z",
            "results": {
                key: {
                    "status": self.results[key].status,
                    "details": self.results[key].details,
                }
                for key in REPORT_FIELDS
            },
            "legal_urls": self.legal_urls,
            "screenshots": self.screenshots,
            "overall_status": "PASS"
            if all(self.results[key].status == "PASS" for key in REPORT_FIELDS)
            else "FAIL",
        }
        return report


def _create_page(playwright: Playwright, args: argparse.Namespace) -> tuple[Page, object]:
    if args.cdp_endpoint:
        browser = playwright.chromium.connect_over_cdp(args.cdp_endpoint)
        if browser.contexts:
            context = browser.contexts[0]
        else:
            context = browser.new_context()
        page = context.pages[0] if context.pages else context.new_page()
        return page, browser

    browser = playwright.chromium.launch(headless=not args.headed)
    context = browser.new_context(viewport={"width": 1440, "height": 900})
    page = context.new_page()
    return page, browser


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="SaleADS Mi Negocio full workflow validation")
    parser.add_argument(
        "--start-url",
        default=None,
        help="Login page URL for the active SaleADS environment. Not hardcoded by this test.",
    )
    parser.add_argument(
        "--cdp-endpoint",
        default=None,
        help="Optional CDP endpoint to attach to an already-open browser session.",
    )
    parser.add_argument(
        "--account-email",
        default="juanlucasbarbiergarzon@gmail.com",
        help="Google account email to select when account chooser appears.",
    )
    parser.add_argument(
        "--artifacts-dir",
        default="artifacts/saleads_mi_negocio_full_test",
        help="Directory where screenshots and final report are written.",
    )
    parser.add_argument(
        "--headed",
        action="store_true",
        help="Run with visible browser window (default is headless).",
    )
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    timestamp = datetime.utcnow().strftime("%Y%m%d_%H%M%S")
    artifacts_dir = Path(args.artifacts_dir) / timestamp
    artifacts_dir.mkdir(parents=True, exist_ok=True)

    with sync_playwright() as playwright:
        page, browser = _create_page(playwright, args)
        try:
            if args.start_url:
                page.goto(args.start_url, wait_until="domcontentloaded", timeout=45000)
            elif page.url in ("about:blank", "data:,"):
                raise RuntimeError(
                    "No start URL was provided and no preloaded page was available. "
                    "Use --start-url or --cdp-endpoint connected to an open SaleADS login page."
                )

            workflow = SaleadsMiNegocioWorkflow(
                page=page,
                artifacts_dir=artifacts_dir,
                account_email=args.account_email,
            )
            final_report = workflow.run_all()

            report_path = artifacts_dir / "final_report.json"
            report_path.write_text(json.dumps(final_report, ensure_ascii=False, indent=2), encoding="utf-8")
            print(json.dumps(final_report, ensure_ascii=False, indent=2))
            print(f"\nFinal report path: {report_path}")

            return 0 if final_report["overall_status"] == "PASS" else 1
        finally:
            browser.close()


if __name__ == "__main__":
    sys.exit(main())

File: qa/saleads-e2e/saleads_mi_negocio_full_test.mjs
import fs from "node:fs/promises";
import path from "node:path";
import { chromium } from "playwright";

const TEST_NAME = "saleads_mi_negocio_full_test";
const GOOGLE_ACCOUNT = process.env.SALEADS_GOOGLE_ACCOUNT ?? "juanlucasbarbiergarzon@gmail.com";
const LOGIN_URL = process.env.SALEADS_LOGIN_URL ?? process.env.SALEADS_URL ?? "";
const HEADLESS = process.env.HEADLESS !== "false";
const SLOW_MO = Number(process.env.PW_SLOW_MO ?? 0);
const NOW = new Date().toISOString().replace(/[:.]/g, "-");
const ARTIFACTS_DIR = path.resolve("artifacts", `${TEST_NAME}-${NOW}`);
const REPORT_PATH = path.join(ARTIFACTS_DIR, "report.json");

const report = {
  name: TEST_NAME,
  startedAt: new Date().toISOString(),
  config: {
    loginUrlProvided: Boolean(LOGIN_URL),
    headless: HEADLESS,
    slowMo: SLOW_MO,
    googleAccount: GOOGLE_ACCOUNT
  },
  evidence: [],
  stepResults: {},
  legalUrls: {
    terminosYCondiciones: null,
    politicaDePrivacidad: null
  }
};

const orderedStepNames = [
  "Login",
  "Mi Negocio menu",
  "Agregar Negocio modal",
  "Administrar Negocios view",
  "Información General",
  "Detalles de la Cuenta",
  "Tus Negocios",
  "Términos y Condiciones",
  "Política de Privacidad"
];

function setStepResult(stepName, status, details) {
  report.stepResults[stepName] = { status, details };
}

function markBlockedStepsAfterLogin() {
  for (const stepName of orderedStepNames.slice(1)) {
    if (!report.stepResults[stepName]) {
      setStepResult(stepName, "FAIL", "Blocked because Login step did not complete.");
    }
  }
}

async function ensureDir(dirPath) {
  await fs.mkdir(dirPath, { recursive: true });
}

async function waitForUi(page) {
  await page.waitForLoadState("domcontentloaded", { timeout: 15000 }).catch(() => {});
  await page.waitForLoadState("networkidle", { timeout: 15000 }).catch(() => {});
  await page.waitForTimeout(700);
}

async function resolveFirstVisible(candidates, timeoutMs, description) {
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    for (const candidate of candidates) {
      const located = candidate.first();
      const count = await candidate.count().catch(() => 0);
      if (count > 0 && (await located.isVisible().catch(() => false))) {
        return located;
      }
    }
    await new Promise((resolve) => setTimeout(resolve, 250));
  }

  throw new Error(`Could not find visible element for: ${description}`);
}

async function expectVisible(candidates, timeoutMs, description) {
  const located = await resolveFirstVisible(candidates, timeoutMs, description);
  await located.waitFor({ state: "visible", timeout: timeoutMs });
  return located;
}

async function clickAndWait(page, locator) {
  await locator.click();
  await waitForUi(page);
}

async function screenshot(page, fileName, fullPage = false) {
  const outputPath = path.join(ARTIFACTS_DIR, fileName);
  await page.screenshot({ path: outputPath, fullPage });
  report.evidence.push(outputPath);
}

async function validateLegalPage({
  page,
  context,
  linkText,
  headingPattern,
  screenshotName,
  urlKey
}) {
  const link = await expectVisible(
    [page.getByRole("link", { name: new RegExp(linkText, "i") }), page.getByText(new RegExp(linkText, "i"))],
    20000,
    linkText
  );

  const previousUrl = page.url();
  const popupPromise = context.waitForEvent("page", { timeout: 12000 }).catch(() => null);
  const navPromise = page.waitForNavigation({ timeout: 12000 }).catch(() => null);

  await link.click();
  await waitForUi(page);

  const popup = await popupPromise;
  await navPromise.catch(() => {});

  let targetPage = page;
  if (popup) {
    targetPage = popup;
    await waitForUi(targetPage);
  }

  await expectVisible(
    [targetPage.getByRole("heading", { name: headingPattern }), targetPage.getByText(headingPattern)],
    20000,
    `Heading ${headingPattern}`
  );

  const pageBody = await targetPage.locator("body").innerText().catch(() => "");
  if (pageBody.trim().length < 120) {
    throw new Error(`Legal content appears too short for ${linkText}.`);
  }

  report.legalUrls[urlKey] = targetPage.url();
  await screenshot(targetPage, screenshotName, true);

  if (popup) {
    await targetPage.close().catch(() => {});
    await page.bringToFront();
    await waitForUi(page);
  } else if (page.url() !== previousUrl) {
    await page.goBack({ waitUntil: "domcontentloaded" }).catch(() => {});
    await waitForUi(page);
  }
}

async function writeReportAndExit() {
  const summary = {};
  for (const stepName of orderedStepNames) {
    summary[stepName] = report.stepResults[stepName]?.status ?? "FAIL";
  }

  report.summary = summary;
  report.finishedAt = new Date().toISOString();
  report.overallStatus = Object.values(summary).every((value) => value === "PASS") ? "PASS" : "FAIL";

  await fs.writeFile(REPORT_PATH, JSON.stringify(report, null, 2), "utf-8");
  console.log(JSON.stringify(report, null, 2));

  process.exitCode = report.overallStatus === "PASS" ? 0 : 1;
}

async function run() {
  await ensureDir(ARTIFACTS_DIR);

  const browser = await chromium.launch({ headless: HEADLESS, slowMo: SLOW_MO });
  const context = await browser.newContext({ viewport: { width: 1600, height: 1000 } });
  const page = await context.newPage();

  let loginSucceeded = false;

  try {
    if (!LOGIN_URL) {
      throw new Error(
        "Set SALEADS_LOGIN_URL to the current environment login page. The test is URL-agnostic and does not hard-code a domain."
      );
    }

    await page.goto(LOGIN_URL, { waitUntil: "domcontentloaded" });
    await waitForUi(page);

    const loginButton = await expectVisible(
      [
        page.getByRole("button", { name: /sign in with google|continue with google|continuar con google/i }),
        page.getByRole("link", { name: /sign in with google|continue with google|continuar con google/i }),
        page.getByText(/sign in with google|continue with google|continuar con google/i)
      ],
      30000,
      "Sign in with Google button"
    );

    const popupPromise = context.waitForEvent("page", { timeout: 10000 }).catch(() => null);
    await clickAndWait(page, loginButton);
    const googlePage = (await popupPromise) ?? page;

    await waitForUi(googlePage);

    const accountChoice = await resolveFirstVisible(
      [googlePage.getByText(GOOGLE_ACCOUNT, { exact: true }), googlePage.getByText(new RegExp(GOOGLE_ACCOUNT, "i"))],
      12000,
      `Google account ${GOOGLE_ACCOUNT}`
    ).catch(() => null);

    if (accountChoice) {
      await clickAndWait(googlePage, accountChoice);
    }

    if (googlePage !== page) {
      await googlePage.waitForEvent("close", { timeout: 70000 }).catch(() => {});
      await page.bringToFront();
    }

    await waitForUi(page);

    await expectVisible([page.locator("aside"), page.locator("[role='navigation']"), page.locator("nav")], 30000, "Left sidebar");
    await expectVisible([page.getByText(/mi negocio/i), page.getByText(/negocio/i)], 30000, "Main application sidebar content");

    await screenshot(page, "01-dashboard-loaded.png", true);
    setStepResult("Login", "PASS", "Main app and left sidebar are visible after Google login.");
    loginSucceeded = true;
  } catch (error) {
    setStepResult("Login", "FAIL", error instanceof Error ? error.message : String(error));
    await screenshot(page, "01-login-failed.png", true).catch(() => {});
  }

  if (!loginSucceeded) {
    markBlockedStepsAfterLogin();
    await context.close();
    await browser.close();
    await writeReportAndExit();
    return;
  }

  try {
    const negocioSection = await expectVisible(
      [page.getByRole("button", { name: /negocio/i }), page.getByText(/negocio/i)],
      20000,
      "Negocio section"
    );
    await clickAndWait(page, negocioSection);

    const miNegocioOption = await expectVisible(
      [page.getByRole("button", { name: /mi negocio/i }), page.getByText(/mi negocio/i)],
      20000,
      "Mi Negocio option"
    );
    await clickAndWait(page, miNegocioOption);

    await expectVisible([page.getByText(/agregar negocio/i)], 15000, "Agregar Negocio");
    await expectVisible([page.getByText(/administrar negocios/i)], 15000, "Administrar Negocios");

    await screenshot(page, "02-mi-negocio-expanded-menu.png", true);
    setStepResult("Mi Negocio menu", "PASS", "Mi Negocio submenu expanded with both required options.");
  } catch (error) {
    setStepResult("Mi Negocio menu", "FAIL", error instanceof Error ? error.message : String(error));
  }

  try {
    const agregarNegocio = await expectVisible([page.getByText(/agregar negocio/i)], 20000, "Agregar Negocio menu item");
    await clickAndWait(page, agregarNegocio);

    await expectVisible([page.getByText(/crear nuevo negocio/i)], 20000, "Crear Nuevo Negocio title");
    const nombreInput = await expectVisible(
      [page.getByLabel(/nombre del negocio/i), page.getByPlaceholder(/nombre del negocio/i), page.locator("input")],
      20000,
      "Nombre del Negocio input"
    );
    await expectVisible([page.getByText(/tienes 2 de 3 negocios/i)], 15000, "Tienes 2 de 3 negocios");
    await expectVisible([page.getByRole("button", { name: /cancelar/i })], 15000, "Cancelar button");
    await expectVisible([page.getByRole("button", { name: /crear negocio/i })], 15000, "Crear Negocio button");

    await screenshot(page, "03-agregar-negocio-modal.png", true);

    await nombreInput.click();
    await nombreInput.fill("Negocio Prueba Automatización");
    const cancelButton = await expectVisible([page.getByRole("button", { name: /cancelar/i })], 15000, "Cancelar button");
    await clickAndWait(page, cancelButton);

    setStepResult("Agregar Negocio modal", "PASS", "Modal validations passed and optional input/cancel actions executed.");
  } catch (error) {
    setStepResult("Agregar Negocio modal", "FAIL", error instanceof Error ? error.message : String(error));
  }

  try {
    const miNegocioOption = await expectVisible(
      [page.getByRole("button", { name: /mi negocio/i }), page.getByText(/mi negocio/i)],
      20000,
      "Mi Negocio option (re-open if collapsed)"
    );
    await clickAndWait(page, miNegocioOption);

    const administrarNegocios = await expectVisible(
      [page.getByText(/administrar negocios/i), page.getByRole("button", { name: /administrar negocios/i })],
      20000,
      "Administrar Negocios"
    );
    await clickAndWait(page, administrarNegocios);

    await expectVisible([page.getByText(/informaci[oó]n general/i)], 20000, "Información General section");
    await expectVisible([page.getByText(/detalles de la cuenta/i)], 20000, "Detalles de la Cuenta section");
    await expectVisible([page.getByText(/tus negocios/i)], 20000, "Tus Negocios section");
    await expectVisible([page.getByText(/secci[oó]n legal/i)], 20000, "Sección Legal section");

    await screenshot(page, "04-administrar-negocios-account-page.png", true);
    setStepResult("Administrar Negocios view", "PASS", "Account page loaded with all required sections.");
  } catch (error) {
    setStepResult("Administrar Negocios view", "FAIL", error instanceof Error ? error.message : String(error));
  }

  try {
    await expectVisible([page.getByText(/business plan/i)], 20000, "BUSINESS PLAN");
    await expectVisible([page.getByRole("button", { name: /cambiar plan/i }), page.getByText(/cambiar plan/i)], 20000, "Cambiar Plan");

    const bodyText = await page.locator("body").innerText().catch(() => "");
    if (!bodyText.match(/@/)) {
      throw new Error("Could not detect user email in visible page text.");
    }

    setStepResult("Información General", "PASS", "User details, BUSINESS PLAN, and Cambiar Plan are visible.");
  } catch (error) {
    setStepResult("Información General", "FAIL", error instanceof Error ? error.message : String(error));
  }

  try {
    await expectVisible([page.getByText(/cuenta creada/i)], 15000, "Cuenta creada");
    await expectVisible([page.getByText(/estado activo/i), page.getByText(/activo/i)], 15000, "Estado activo");
    await expectVisible([page.getByText(/idioma seleccionado/i), page.getByText(/idioma/i)], 15000, "Idioma seleccionado");
    setStepResult("Detalles de la Cuenta", "PASS", "All required account detail fields are visible.");
  } catch (error) {
    setStepResult("Detalles de la Cuenta", "FAIL", error instanceof Error ? error.message : String(error));
  }

  try {
    await expectVisible([page.getByText(/tus negocios/i)], 15000, "Tus Negocios section");
    await expectVisible(
      [page.getByRole("button", { name: /agregar negocio/i }), page.getByText(/agregar negocio/i)],
      15000,
      "Agregar Negocio button in business section"
    );
    await expectVisible([page.getByText(/tienes 2 de 3 negocios/i)], 15000, "Tienes 2 de 3 negocios in business section");
    setStepResult("Tus Negocios", "PASS", "Business list and limits are visible.");
  } catch (error) {
    setStepResult("Tus Negocios", "FAIL", error instanceof Error ? error.message : String(error));
  }

  try {
    await validateLegalPage({
      page,
      context,
      linkText: "Términos y Condiciones",
      headingPattern: /t[eé]rminos y condiciones/i,
      screenshotName: "05-terminos-y-condiciones.png",
      urlKey: "terminosYCondiciones"
    });
    setStepResult("Términos y Condiciones", "PASS", "Legal page heading/content validated and URL captured.");
  } catch (error) {
    setStepResult("Términos y Condiciones", "FAIL", error instanceof Error ? error.message : String(error));
  }

  try {
    await validateLegalPage({
      page,
      context,
      linkText: "Política de Privacidad",
      headingPattern: /pol[ií]tica de privacidad/i,
      screenshotName: "06-politica-de-privacidad.png",
      urlKey: "politicaDePrivacidad"
    });
    setStepResult("Política de Privacidad", "PASS", "Legal page heading/content validated and URL captured.");
  } catch (error) {
    setStepResult("Política de Privacidad", "FAIL", error instanceof Error ? error.message : String(error));
  }

  await context.close();
  await browser.close();
  await writeReportAndExit();
}

run().catch(async (error) => {
  setStepResult("Login", "FAIL", `Unexpected script error: ${error instanceof Error ? error.message : String(error)}`);
  markBlockedStepsAfterLogin();
  await writeReportAndExit();
});
